File: simulation_v7.R
#######################################
# Authors: Hannah Martiny & Chris Workman
# DTU Biobuilders project for iGEM 2018
#######################################

library(stats)
library(scales)
library(RColorBrewer)
library(ggplot2)
library(reshape2)

setwd("~/Google\ Drev/igem/")
#setwd("~/Google Drive/Biobuilders/iGEM 2018/Dry lab/Mycelium Growth/simulation/")

source("functions_v1.R")
source("IntersectLineSeg.R")
#source("letters/letter_funcs.R")

#######################################
#hyCols = rev(brewer.pal(9, "PuBu"))
# hyCols = rev(brewer.pal(9, "OrRd"))
# hyCols = rev(brewer.pal(9, "BuGn"))
# hyCols = rev(brewer.pal(9, "PuRd"))
# hyCols = rev(brewer.pal(9, "PuRd"))
hyCols = rev(brewer.pal(9, "YlGnBu")[3:9])

# PARAMETERS USED IN THE SIMULATION
leftright = c(-1, 1)    # branching left or right
minTheta = 10           #*pi/180 - minimum angle a branch can occur
maxTheta = 100          #*pi/180 - maximum angle a branch can occur

avgRate = 50            # average tip extension rate (?m/h), source: Spohr et al. 1998
q  = 0.005              # branching frequency (maybe need to be scaled of what the time step is)
N  = 400               # max simulation rounds
M  = 1e5                # max hyphae
tstep = 0.0005          # time step (h)

ktip1 = 80              # initial tip extension rate, value estimated from Spohr et al 1998 figure 5
maxktip = 155           # maximum tip extension rate, value estimated from Spohr et al 1998 figure 5
ktip2 = maxktip - ktip1 # difference between ktip1 and maxktip
Kt = 5                  # saturation constant

###########################################
# Single square as growth area
# put init_n spores on centers and define maxium width
centers = matrix(c(0,0), 1, 2)
width = 50
init_n = 10

# lists for saving results during the simulation
snapShots = list()
hyphae <- list()
hyphaeSnapshots <- list()
# initializes the first spore(s) and growth angles
hNum = 1
for(ii in 1:dim(centers)[1]) {
  for(jj in 1:init_n) {
    rxy = runif(2, -width, width) + as.integer(centers[ii,])
    iTheta = round(runif(min=0, max=360, 1))
    hyphae[[hNum]] <- c(x0=rxy[1], y0=rxy[2], x=rxy[1], y=rxy[2], 
                        angle=iTheta, nevents=0, t=0, l=0) ## Initial spores
    hNum = hNum + 1
  }
}

# save initial state of simulation and plot it
snapShots[["0"]] = data.frame(do.call(rbind, hyphae))
hyphaeSnapshots[["0"]] = hyphae
par(mfrow=c(1,1))
plot_hyphae(x=snapShots[["0"]], tMax=100, hcols=hyCols, xlim=c(-100,100), ylim=c(-100,100))


S0 = 50000            # initial substrate concentration
St = array(S0, N)      # track total substrate level over time
S_snapshots = list()
# substrate grid setup
S_bbs = grid_bounding_boxes(w=10, xrng=c(-70, 70), yrng=c(-70, 70))
S <- array(S0/dim(S_bbs)[1], dim(S_bbs)[1])
S_snapshots[["0"]] = data.frame(S)





dls = array(0, N)     # track biomass densities over time
r = 1                 # forgot specific name, but used to scale biomass density

# actual simulation
i = 0; m = 0
while (i <= N & m < M) {
  i = i + 1
  m = length(hyphae) # number of hyphae at this step
  
  # get where each hyphae tip is in the substrate grid
  hyphae_substrate = hyphae_hits_substrate(hl = hyphae, bbs = S_bbs)

  dl = 0 # track biomass at this step
  
  # loop through hyphae, find new coordinates for tips
  for (j in 1:m) {
    # check if enough substrate around substrate tip
    if (sum(hyphae_substrate[j, ])) {
      angle = hyphae[[j]]["angle"][[1]]
      hi = as.numeric(hyphae[[j]][1:4])
      
      # tip extension vlaue, calculated as the accelerated growth * substrate limitation 
      # equation of the Monod type
      extension = tipExtensionMonod(ktip1 = ktip1, ktip2 = ktip2, Kt = Kt, l = hyphae[[j]]["l"][[1]], S=S[hyphae_substrate[j, ] == 1], Ks=200)
      dx = extension * tstep * cos(angle*pi/180) # new coordinate in x-axis
      dy = extension * tstep * sin(angle*pi/180) # new coordinate in y-axis
      # biomass just created for hyphae j
      dl_c = sqrt(dx^2 + dy^2)
      dl = dl + dl_c
      
      # substrate used by the hyphal tip
      S[hyphae_substrate[j, ] == 1] = S[hyphae_substrate[j, ] == 1] - min(r*dl_c, S[hyphae_substrate[j, ] == 1])
      
      # update data frame
      hyphae[[j]]["x"] = hi[3] + dx
      hyphae[[j]]["y"] = hi[4] + dy
      hyphae[[j]]["l"] = hyphal_length(hyphae[[j]])
      b = hyphae[[j]]["nevents"]
    
      qApl = q
      #qApl = q/(b+1)
      
      if (runif(1.0) < qApl) { # branching event (q should depend on l and/or nevents)
        # the direcction a new branch will grow
        newdir = sample(leftright, 1)
        newangle = angle + newdir*round( runif(min=minTheta, max=maxTheta, 1) )
  
        # nevents goes one up, as there have been branching event for this hyphae
        hyphae[[j]]["nevents"] = hyphae[[j]]["nevents"] + 1
  
        # add new hyphae
        hyphae[[length(hyphae) + 1]] <- c(x0 = hi[3], y0 = hi[4],
                                          x  = hi[3], y  = hi[4],
                                          angle=newangle, nevents=0, t=i, l=0)
      }
    }
  }
  ## Subtract the resources consumed in the last round of growth
  # S[i+1] = S[i] - min(r*dl, S[i])
  St[i+1] = sum(S)
  
  # save biomass at this timestep
  dls[i+1] = dl


  if(i %% 10 == 0) {
    print(paste("Iteration =", i, "hNum =", length(hyphae) ))
    snapShots[[as.character(i)]] = data.frame(do.call(rbind, hyphae))
    S_snapshots[[as.character(i)]] = data.frame(S)
    hyphaeSnapshots[[as.character(i)]] = hyphae
  }
}

# create data structures and x-y limits used for plotting
m = length(hyphae)
dat = data.frame(do.call(rbind, hyphae))
x_range <- range(dat[, c("x", "x0")])
y_range <- range(dat[, c("y", "y0")])
tMax = max(dat[,"t"])

# plot(as.matrix(cbind(as.integer(names(extensions))*tstep, extensions)), ylab = "average extension rate")

################################################
names(snapShots)
################################################

# Simple gallery
pdf("example_n10_v4_S010000.pdf", useDingbats=FALSE)
par(mfrow=c(3, 4), mar=c(2,2,3,1))
for(i in names(snapShots)[seq(2, length(snapShots), 2)]) {
  sst = sprintf("%.3f", as.integer(i)*tstep)
  plot_hyphae(x=snapShots[[i]], tMax=max(dat[,"t"]), hcols=hyCols, xlim=x_range, ylim=y_range, main=paste("Time =", sst))
}
plot_hyphae(x=dat, tMax=max(dat[,"t"]), hcols=hyCols, xlim=x_range, ylim=y_range, main=paste("Time =", max(dat[,"t"])*tstep))

dev.off()

# substrate change recorded at the same time as the snapshots data
# therefore visualize it
pdf("example_n10_substrate_tracked_S010000.pdf", useDingbats=FALSE)
par(mfrow=c(3, 4), mar=c(2,2,3,1))
for (i in names(S_snapshots)[seq(2, length(S_snapshots), 2)]){
  sst = sprintf("%.3f", as.integer(i)*tstep)
  plot_hyphae(x=snapShots[[i]], tMax=max(dat[,"t"]), hcols=hyCols, xlim=x_range, ylim=y_range, main=paste("Time =", sst))
  scpal = alpha(colorRampPalette(brewer.pal(9, "OrRd"))(round(1.1*max(S_snapshots[[i]]))), 0.5)
  for(j in 1:dim(S_bbs)[1]){
    polygon(x=S_bbs[j,c(1,1,3,3)], y=S_bbs[j,c(2,4,4,2)], lty=1, col=scpal[round(S_snapshots[[i]][j,])+1], border=NA)
  }
}
dev.off()
# plot(NA, type='n', xlim=x_range, ylim=y_range, xlab="", ylab="", xaxt='n', yaxt='n', bty='n')


################################################
# Animation of the mycelium growth
# Create a MANY png's and then combine them into one gif
if(TRUE) {
  for(i in names(snapShots)) {
    sst = sprintf("%.3f", as.integer(i)*tstep)
    #print(sst)
    # HYPHAL GROWTH PLOT
    fname = paste("ani/Animate_hyphal_", sprintf("%04d", as.integer(i)) ,".png", sep="")
    print(fname)
    png(file=fname, width=800, height=800)#, bg = "transparent")
    par(mar=c(2,2,3,1))
    #plot_hyphae(x=snapShots[[i]], tMax=max(dat[,"t"]), hcols=hyCols, main=paste("Time =", sst))
    plot_hyphae(x=snapShots[[i]], tMax=max(dat[,"t"]), hcols=hyCols, xlim=x_range, ylim=y_range, main="", bty='n', yaxt='n',  xaxt='n', ann=FALSE)
    dev.off()
    
    # SUBSTRATE GROWTH PLOT
    fname = paste("ani/Animate_substrate_convergence_", sprintf("%04d", as.integer(i)) ,".png", sep="")
    print(fname)
    png(file=fname, width=800, height=800)#, bg = "transparent")
    par(mar=c(2,2,3,1))
    #plot_hyphae(x=snapShots[[i]], tMax=max(dat[,"t"]), hcols=hyCols, main=paste("Time =", sst))
    plot_hyphae(x=snapShots[[i]], tMax=max(dat[,"t"]), hcols=hyCols, xlim=x_range, ylim=y_range, main="", bty='n', yaxt='n',  xaxt='n', ann=FALSE)
    scpal = alpha(colorRampPalette(brewer.pal(9, "OrRd"))(round(1.1*max(S_snapshots[[i]]))), 0.5)
    for(j in 1:dim(S_bbs)[1]){
      polygon(x=S_bbs[j,c(1,1,3,3)], y=S_bbs[j,c(2,4,4,2)], lty=1, col=scpal[round(S_snapshots[[i]][j,])+1], border=NA)
    }
    dev.off()
  }
  setwd("ani")
  cmd = sprintf("convert *.png -delay 12 -loop 0 Simulation.gif\n", init_n)
  system(cmd)
  #system("rm *Animate_*.png\n")
  setwd("..")
}

################################################
# Make summary of total hyphal lengths and tip count at time points
growthSummary = data.frame(hyphae = unlist(lapply(snapShots, nrow)), total_l=NA)
for(i in rownames(growthSummary)){
  growthSummary[i,"total_l"] = sum(snapShots[[i]][,"l"])
}


# Boring views...
# Maybe better with ggplot2? (probably)
# pdf(file="Summaries_n10_FINAL.pdf", useDingbats = FALSE)
# par(mfrow=c(2,2))
png("hyphal_length_hours.png")
hours = as.integer(rownames(growthSummary))*tstep
plot(hours, growthSummary$total_l, pch=20, ylim=c(5, max(growthSummary$total_l, growthSummary$hyphae)),
     log="y", type='b', xlab="Hours", ylab="Length (um), Tip number")
points(hours, growthSummary$hyphae, pch=21, col='red', type='b')
legend("topleft", legend=c("Total hyphal length", "Hyphal tips"), fill=c("black", "red"))
dev.off()

png("hyphal_length_hist.png")
hist(dat[,"l"], xlab="Length (um)", main="Hyphal length")
dev.off()

png("branches_hyphae_hist.png")
hist(dat[,"nevents"], xlab="Branches", main="Number of branches per hyphae")
dev.off()

# some more views
# plots of substrate and biomass changes during simulation

# par(mfrow=c(2, 1))
png("substrate_level.png")
plot(1:tMax*tstep, St[1:tMax], type='l', ylim = c(0, max(St)), xlab = "Hours", ylab = "Substrate")
dev.off()
png("biomass_level.png")
dl_max = max(dls)
dls[c(which(max(dls)==dls)+1:length(dls))] = dl_max
plot(1:tMax*tstep, dls[1:tMax], type='l', ylim = c(0, max(dls, na.rm=T)), xlab = "Hours", ylab = "Biomass")
dev.off()
################################################
# Density calculations
w = 10
xrng = c(-70, 70)
yrng = c(-70, 70)
bbs = grid_bounding_boxes(w=10, xrng=xrng, yrng=yrng)
(n = dim(bbs)[1])
hyphae2boxes = hyphae_hits(hl=hyphae, bbs=bbs)
d = hyphal_length_by_RAE(hl=hyphae, h2bbs=hyphae2boxes, bbs=bbs, plotting=FALSE)
round(100*(sum(d)/sum(dat[,"l"])),2) ## % covered
#nPerBox = colSums(hyphae2boxes)

# Inspect the process for a sample of BBs
#rinds = 70; xy = bbs[rinds,]
#plot(NA, type='n', xlim=c(xy[1]-w, xy[1]+w), ylim=c(xy[2]-w, xy[2]+w))
#rinds = sample(1:n, 20)
#plot(NA, type='n', xlim=c(-100, 100), ylim=c(-100,100))
##for(i in 1:dim(bbs)[1]) polygon(x=bbs[i,c(1,1,3,3)], y=bbs[i,c(2,4,4,2)], lty=3)
#dsamp = hyphal_length_by_RAE(hl=hyphae, h2bbs=hyphae2boxes[,rinds, drop=FALSE], bbs=bbs[rinds,,drop=FALSE], plotting=TRUE)


################################################
hist(d, breaks=seq(0, max(d+10),10))

################################################
# Plot the density like a heatmap
range(d, na.rm=TRUE)
cpal = alpha(colorRampPalette(brewer.pal(9, "Reds"))(round(1.1*max(d, na.rm=T))), 0.5)
par(mfrow=c(1,1))
xyrng = c(-80,80)

pdf(file="density_map_n10_FINAL.pdf", useDingbats = FALSE)
plot_hyphae(x=dat, tMax=max(dat[,"t"]), hcols=hyCols, xlim=x_range, ylim=y_range, main=paste("Time =", max(dat[,"t"])*tstep))
#plot(NA, type='n', xlim=x_range, ylim=y_range, xlab="", ylab="", xaxt='n', yaxt='n', bty='n')
for(j in 1:n){
  polygon(x=bbs[j,c(1,1,3,3)], y=bbs[j,c(2,4,4,2)], lty=1, col=cpal[round(d[j])+1], border=NA)
}
dev.off()

################################################
# make animation of density heatmaps
for (i in names(snapShots)) {
  snapHyphae = hyphaeSnapshots[[i]]
  hyphae2boxes = hyphae_hits(hl=snapHyphae, bbs=bbs)
  d = hyphal_length_by_RAE(hl=snapHyphae, h2bbs=hyphae2boxes, bbs=bbs, plotting=FALSE)
  d[is.na(d)] = 0
  cpal = alpha(colorRampPalette(brewer.pal(9, "Reds"))(round(1.1*max(d))), 0.5)
  
  fname = paste("ani/Animate_density_", sprintf("%04d", as.integer(i)) ,".png", sep="")
  print(fname)
  png(file=fname, width=800, height=800)#, bg = "transparent")
  par(mar=c(2,2,3,1))
  plot_hyphae(x=snapShots[[i]], tMax=max(dat[,"t"]), hcols=hyCols, xlim=x_range, ylim=y_range, main="", bty='n', yaxt='n',  xaxt='n', ann=FALSE)
  for(j in 1:n){
    polygon(x=bbs[j,c(1,1,3,3)], y=bbs[j,c(2,4,4,2)], lty=1, col=cpal[round(d[j])+1], border=NA)
  }
  dev.off()
}

################################################
# output density to file for Mathias
d_matrix <- matrix(d, nrow = 5, ncol=5)
write.csv(x = d_matrix, 'PDE_initial_density.csv', col.names = FALSE, row.names=FALSE)

File: simulation_v7.py
#!/usr/bin/env python3
# coding=utf-8
import sys
from datetime import datetime
import argparse
import pandas as pd
import numpy as np
import math
import matplotlib.pyplot as plt
from matplotlib import collections as mc
from matplotlib import cm
import copy
import pdb
from matplotlib.backends.backend_pdf import PdfPages
import matplotlib
from matplotlib import colors
from matplotlib.patches import Rectangle
from matplotlib.collections import PatchCollection
matplotlib.rcParams.update({'font.size': 6})

def get_parser():
    # build commandline parser
    parser = argparse.ArgumentParser(
        description="Script for simulating hyphal growth.")

    # arguments currently not in use - parameters are hardcoded in the script
    # arguments could easily be integrated for easier tests of parameter changes (in bash)
    # I have included all parameters that we can adjust
    """
    parser.add_argument("-minTheta", type=int, dest="minTheta", default=10,
                        help="*pi/180 - minimum angle a branch can occur. Default: 10")
    parser.add_argument("-maxTheta", type=int, dest="maxTheta", default=100,
                        help="*pi/180 - maximum angle a branch can occur. Default: 100")
    parser.add_argument("-avgRate", type=int, dest="avgRate", default=50,
                        help="average tip extension rate (?m/h), source: Spohr et al. 1998. Default: 50")
    parser.add_argument("-q", type=float, dest="q", default=0.005,
                        help="branching frequency (maybe need to be scaled of what the time step is). Default: 0.005")
    parser.add_argument("-N", type=int, dest="N", default=1000,
                        help="max simulation rounds. Default: 1000")
    parser.add_argument("-M", type=int, dest="M", default=100000,
                        help="Max hyphae (length?). Default: 100000")
    parser.add_argument("-tstep", type=float, dest="tstep", default=0.0005,
                        help="Time step (h). Default: 0.0005")
    parser.add_argument("-ktip1", type=int, dest="ktip1", default=80,
                        help="Initial tip extension rate, value estimated from Spohr et al 1998 figure 5. Default: 80")
    parser.add_argument("-maxktip", type=int, dest="maxktip", default=155,
                        help="Maximum tip extension rate, value estimated from Spohr et al 1998 figure 5. Default: ")
    parser.add_argument("-Kt", type=int, dest="Kt", default=5,
                        help="Saturation constant. Default: 5")
    """
    return parser

def get_args():
    parser = get_parser()
    args = parser.parse_args()
    return args

def plot_initial_hyphae(hyphae, tMax=100, xlim=(-100,100), ylim=(-100,100)):
    """In R, you get the points if coordinates connecting line segments are identical when doing normal scatterplot.
    But with pyplot you don't - therefore this function for plotting initial hyphae"""
    fig, ax = plt.subplots(figsize=(5, 5))
    points = list(zip(hyphae['x0'],hyphae['y0']))
    plt.plot(points, ',', color='black')
    plt.show()

def plot_hyphae(hyphae, tMax=100, xlim=(-100,100), ylim=(-100,100)):
    """Works as function in R-script, but currently not in use"""
    fig, ax = plt.subplots(figsize=(5, 5))
    line_segments = [[list(zip(hyphae['x0'],hyphae['y0']))[i], list(zip(hyphae['x'],hyphae['y']))[i]] for i in range(len(hyphae))]
    lc = mc.LineCollection(line_segments, linewidths=2)
    ax.add_collection(lc)
    ax.autoscale()
    ax.margins(0.1)
    plt.show()

def grid_bounding_boxes(w=10, xrng=(-70, 70), yrng=(-70, 70)):
    """Function making the same grid_bounding_boxes as in R-script"""
    x0 = [no for no in range(xrng[0], xrng[1]+w, w)]
    y0 = [no for no in range(yrng[0], yrng[1]+w, w)]

    xcenters, ycenters = list(), list()
    for i in range(len(x0)):
        xcenters.extend(x0)
        ycenters.extend([y0[i]]*len(y0))
    centers = pd.DataFrame({0:xcenters, 1:ycenters})
    bbs = pd.concat([centers-(w/2), centers.rename(columns={0:2, 1:3})+(w/2)], axis=1)

    return bbs

def grid_bounding_boxes_lower_left(w=10, xrng=(-70, 70), yrng=(-70, 70)):
    """grid_bounding_boxes function for plotting substrate rectangles
    Returns a dataframe of coordinates of all lower left corners"""
    x0 = [no for no in range(xrng[0], xrng[1]+w, w)]
    y0 = [no for no in range(yrng[0], yrng[1]+w, w)]

    xcenters, ycenters = list(), list()
    for i in range(len(x0)):
        xcenters.extend(x0)
        ycenters.extend([y0[i]]*len(y0))
    centers = pd.DataFrame({0:xcenters, 1:ycenters})

    return centers-(w/2)

def hyphae_hits_substrate(hl, bbs):
    m = len(hl)
    h2b = pd.DataFrame(np.zeros((m, len(bbs))))

    for j in range(m):
        x_val = hl['x'].iloc[j]
        y_val = hl['y'].iloc[j]

        xSAT = (x_val <= bbs[2]) & (bbs[0] <= x_val)
        ySAT = (y_val <= bbs[3]) & (bbs[1] <= y_val)
        bbInds = (xSAT & ySAT).index[(xSAT & ySAT) == True]   # index of the True in boolean series
        h2b.iloc[j, bbInds] = 1
    return h2b

def tipExtensionMonod(ktip1, ktip2, Kt, l, S, Ks):
    """source: Lejeune et al 1995, Morphology of Trichoderma reesei QM 9414 in Submerged Cultures"""
    return (ktip1 + ktip2*(l/(l+Kt))) * S/(S+Ks)

def hyphal_length(x):
  return math.sqrt( (x["x"]-x["x0"])**2 + (x["y"]-x["y0"])**2 )

def main(args):
    # PARAMETERS USED IN THE SIMULATION
    leftright = [-1, 1]  # branching left or right
    minTheta = 10  # *pi/180 - minimum angle a branch can occur
    maxTheta = 100  # *pi/180 - maximum angle a branch can occur

    avgRate = 50  # average tip extension rate (?m/h), source: Spohr et al. 1998
    q = 0.005  # branching frequency (maybe need to be scaled of what the time step is)
    N = 1000  # max simulation rounds
    M = 100000  # max hyphae
    tstep = 0.0005  # time step (h)

    ktip1 = 80  # initial tip extension rate, value estimated from Spohr et al 1998 figure 5
    maxktip = 155  # maximum tip extension rate, value estimated from Spohr et al 1998 figure 5
    ktip2 = maxktip - ktip1  # difference between ktip1 and maxktip
    Kt = 5  # saturation constant

    # Single square as growth area
    # put init_n spores on centers and define maxium width
    centers = [[0,0]]
    width = 50
    init_n = 10

    # lists for saving results during the simulation
    snapShots = dict()
    hyphae_dict = {'x0':[],'y0':[], 'x':[],'y':[],'angle':[]}
    hyphaeSnapshots = dict()

    # initializes the first spore(s) and growth angles
    hNum = 0
    for i in range(len(centers)):
        for j in range(init_n):
            rxy = np.random.uniform(-width, width, 2) + [int(no) for no in centers[i]]
            iTheta = int(np.random.uniform(0,360))

            # initialize spores
            hyphae_dict['x0'].append(rxy[0])
            hyphae_dict['y0'].append(rxy[1])
            hyphae_dict['x'].append(rxy[0]+0.01)
            hyphae_dict['y'].append(rxy[1]+0.01)
            hyphae_dict['angle'].append(iTheta)
            hNum += 1

    # initialize spores
    hyphae_dict['nevents'] = [0 for i in range(hNum)]
    hyphae_dict['t'] = [0 for i in range(hNum)]
    hyphae_dict['l'] = [0 for i in range(hNum)]

    hyphae = pd.DataFrame(hyphae_dict)
    # append initial dataframe to snapshot lists
    snapShots[0] = (copy.deepcopy(hyphae))
    hyphaeSnapshots['0'] = (copy.deepcopy(hyphae))
    #plot_initial_hyphae(snapShots[0])

    S0 = 50000                      # initial substrate concentration
    St = [S0 for i in range(N)]     # track total substrate level over time
    S_snapshots = dict()            # substrate snapshots

    # substrate grid setup
    S_bbs = grid_bounding_boxes(w=10, xrng=(-70, 70), yrng=(-70, 70))
    S = pd.DataFrame({'S':[S0/len(S_bbs) for i in range(len(S_bbs))]})
    S_snapshots[0] = S

    dls = [0]*N                     # track biomass densities over time
    r = 1                           # forgot specific name, but used to scale biomass density

    # actual simulation
    i, m = 0, 0
    while i < N and m < M:
        m = len(hyphae)             # number of hyphae at this step

        # get where each hyphae tip is in the substrate grid
        hyphae_substrate = hyphae_hits_substrate(hl=hyphae, bbs=S_bbs)

        dl = 0                      # track biomass at this step

        # loop through hyphae, find new coordinates for tips
        for j in range(m):
            # check if enough substrate around substrate tip (>0)
            if sum(hyphae_substrate.iloc[j,:]):
                angle = hyphae['angle'].iloc[j]
                hi = hyphae[['x0', 'y0', 'x', 'y']].iloc[j,:]

                # tip extension value, calculated as the accelerated growth * substrate limitation
                # equation of the Monod type
                extension = tipExtensionMonod(ktip1=ktip1, ktip2=ktip2, Kt=Kt, l=hyphae['l'].iloc[j],
                                              S=S.loc[hyphae_substrate.iloc[j,:] == 1].iloc[0,0], Ks=200)
                dx = extension * tstep * math.cos(angle * math.pi / 180)    # new coordinate in x-axis
                dy = extension * tstep * math.sin(angle * math.pi / 180)    # new coordinate in y-axis

                # biomass just created for hyphae j
                dl_c = math.sqrt(dx**2 + dy**2)
                dl += dl_c

                # substrate used by the hyphal tip
                S.loc[hyphae_substrate.iloc[j, :] == 1] = S.loc[hyphae_substrate.iloc[j, :] == 1].iloc[0, 0] - min(r*dl_c, S.loc[hyphae_substrate.iloc[j, :] == 1].iloc[0, 0])

                # update data frame
                hyphae['x'].iloc[j] = hi['x'] + dx
                hyphae['y'].iloc[j] = hi['y'] + dy
                hyphae['l'].iloc[j] = hyphal_length(hyphae.iloc[j,:])
                #b = hyphae['nevents'].iloc[j]

                qApl = q
                # qApl = q/(b+1)        # q could be a function of number of events
                # branching event (q should depend on l and/or nevents)

                if np.random.uniform(0,1) < qApl:
                    # the direcction a new branch will grow - randomly left or right
                    newdir = np.random.choice(leftright)
                    newangle = angle + newdir * round(np.random.uniform(minTheta,maxTheta))

                    # nevents goes one up, as there have been branching event for this hyphae
                    hyphae['l'].iloc[j] += 1

                    # add new hyphae
                    hyphae = hyphae.append({'x0': hi['x'], 'y0': hi['y'], 'x': hi['x'], 'y': hi['y'], 'angle': newangle, 'nevents': 0, 't': i, 'l': 0}, ignore_index=True)

        # subtract the resources consumed in the last round of growth
        # S[i+1] = S[i] - min(r*dl, S[i])
        St[i] = S.sum()
        # save biomass at this timestep
        dls[i] = dl

        # increment i before saving time information
        i += 1
        if (i) % 10 == 0:
            print(f"Iteration = {i}, hNum = {len(hyphae)}")
            snapShots[i] = copy.deepcopy(hyphae)
            S_snapshots[i] = copy.deepcopy(S)
            hyphaeSnapshots[str(i)] = copy.deepcopy(hyphae)

    """
    m = len(hyphae)
    dat = copy.deepcopy(hyphae)
    x_range = (dat[['x0', 'x']].min().min(), dat[['x0', 'x']].max().max())
    y_range = (dat[['y0', 'y']].min().min(), dat[['y0', 'y']].max().max())
    tMax = max(dat['t'])
    """

    print('# Simulation done! Plotting...')
    # dimensions for subplots on one page (n-rows and m-cols)
    n, m = 3, 4
    pdf_name = 'simple_gallery_1000.pdf'
    with PdfPages(pdf_name) as pdf:
        # initialize layout for plots
        f, axarr = plt.subplots(n, m, sharex='none', sharey='none')
        arr_ij = [(x, y) for x, y in np.ndindex(axarr.shape)]
        subplots = [axarr[index] for index in arr_ij]

        splot_index = 0

        # iterate through snapshots, selecting every second
        for i in range(1, len(snapShots), 2):
            # select every second snapshot sorted by time
            snapshot_to_plot = snapShots[sorted(snapShots)[i]]
            S_snapshot_to_plot = S_snapshots[sorted(snapShots)[i]]
            line_segments = [[list(zip(snapshot_to_plot['x0'], snapshot_to_plot['y0']))[i], list(zip(snapshot_to_plot['x'], snapshot_to_plot['y']))[i]] for i in
                             range(len(snapshot_to_plot))]

            # create continuous norm for mapping colors to hyphae according to time they occured
            hyphae_norm = plt.Normalize(hyphae['t'].min(), hyphae['t'].max())
            lc = mc.LineCollection(line_segments, linewidths=0.5, cmap='GnBu_r', norm=hyphae_norm)
            lc.set_array(snapshot_to_plot['t'])
            subplots[splot_index].add_collection(lc)

            # create continuous norm for substrate color mapping
            substrate_norm = plt.Normalize(0, S0/len(S_bbs))
            #subplots[splot_index].imshow(np.array(S).reshape(15,15), cmap='YlOrBr_r', norm=substrate_norm)
            S_plot_bbs = grid_bounding_boxes_lower_left(w=10, xrng=(-70, 70), yrng=(-70, 70))

            YlOrBr = cm.get_cmap('YlOrBr')
            patches, colors = list(), list()
            for j in range(len(S_plot_bbs)):
                rect = plt.Rectangle((S_plot_bbs.iloc[j,0], S_plot_bbs.iloc[j,1]), width=10, height=10) #color=YlOrBr(S_snapshot_to_plot.iloc[j]/S0/len(S_bbs))
                patches.append(rect)
            #pdb.set_trace()

            p = PatchCollection(patches, cmap=cm.get_cmap('YlOrBr'), alpha=0.5, norm=substrate_norm)
            p.set_array(S_snapshot_to_plot['S'])
            subplots[splot_index].add_collection(p)

            subplots[splot_index].autoscale()
            subplots[splot_index].margins(0.1)
            subplots[splot_index].set_title(f"Time = {round(sorted(snapShots)[i]*tstep,3)}")
            subplots[splot_index].set_xlim(-60, 60)
            subplots[splot_index].set_ylim(-60, 60)
            subplots[splot_index].set_xticks(list(range(-60,70,30)))
            subplots[splot_index].set_yticks(list(range(-60,70,30)))
            plt.tight_layout()

            # increment index counter
            splot_index += 1
            # a pdf page is full once the count is equal to n*m
            if splot_index == n * m:
                pdf.savefig()
                plt.close(f)
                f, axarr = plt.subplots(n, m, sharex='none', sharey='none')
                arr_ij = [(x, y) for x, y in np.ndindex(axarr.shape)]
                subplots = [axarr[index] for index in arr_ij]
                splot_index = 0

        # save last page
        pdf.savefig()
        plt.close(f)
    return pdf_name


if __name__ == '__main__':
    print("# Running...")
    start_time = datetime.now()
    args = get_args()
    print("# args:", args)
    pdf_name = main(args)
    print(f'# Done! Made pdf: {pdf_name}')
    end_time = datetime.now()
    print('# Duration: {}'.format(end_time - start_time))